File: Leetcode/Oct/23Oct24/23Oct24.js
function dfs(node, depth, parent, nodeInfo) {
    if (!node) return;
    nodeInfo.set(node, { depth: depth, parent: parent });
    dfs(node.left, depth + 1, node, nodeInfo);
    dfs(node.right, depth + 1, node, nodeInfo);
}
function replaceWithCousinSums(root) {
    const nodeInfo = new Map();
    dfs(root, 0, null, nodeInfo);

    const cousinSums = new Map();
    for (const [node, info] of nodeInfo.entries()) {
        const depth = info.depth;
        if (!cousinSums.has(depth)) cousinSums.set(depth, 0);
        cousinSums.set(depth, cousinSums.get(depth) + node.val);
    }

    for (const [node, info] of nodeInfo.entries()) {
        const depth = info.depth;
        const parent = info.parent;
        const cousinSum = cousinSums.get(depth);
        let sum = cousinSum;
        if (parent) {
            if (parent.left) sum -= parent.left.val;
            if (parent.right) sum -= parent.right.val;
        }
        node.val = sum - node.val;
    }
}

File: Leetcode/2024/Oct/23Oct24/23Oct24.py
def dfs(node, depth, parent, node_info):
    if not node:
        return
    node_info[node] = (depth, parent)
    dfs(node.left, depth + 1, node)
    dfs(node.right, depth + 1, node)

node_info = {}
dfs(root, 0, None, node_info)
cousin_sums = {}
for node, (depth, parent) in node_info.items():
    if depth not in cousin_sums:
        cousin_sums[depth] = 0
    cousin_sums[depth] += node.val

for node, (depth, parent) in node_info.items():
    node.val = cousin_sums[depth] - node.val
